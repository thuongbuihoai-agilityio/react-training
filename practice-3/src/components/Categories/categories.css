.categories {
  margin-top: 50px;
  text-align: center;
}

.categories__title {
  text-transform: uppercase;
  font-weight: bold;
}

.categories__list {
  display: flex;
  justify-content: center;
  margin: 30px 0 40px 0;
  padding: 0;
}

.categories__item {
  list-style: none;
  cursor: pointer;
  margin: 0 30px;
  position: relative;
  color: #696969;
}

.categories__item:hover {
  color: #ebb576;
}

.categories__item::after {
  content: "";
  width: 5px;
  height: 5px;
  border-radius: 50%;
  background-color: #555;
  position: absolute;
  right: -30px;
  top: 45%;
  color: #696969;
}

.categories__item > a {
  color: black;
  text-decoration: none;
}