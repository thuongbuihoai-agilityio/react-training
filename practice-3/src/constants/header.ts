const USERS = {
  username: "Hoai Thuong",
};

export { USERS }