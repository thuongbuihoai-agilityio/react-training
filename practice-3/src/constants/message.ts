const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

export { SUCCESS_MSG };
