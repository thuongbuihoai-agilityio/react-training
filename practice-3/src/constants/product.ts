import { Product } from "types/product";

const PRODUCT_CONTENT: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg"
  ]
}

export { PRODUCT_CONTENT };
