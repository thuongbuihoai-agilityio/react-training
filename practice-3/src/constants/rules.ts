const RULES = {
  REQUIRED: "required",
  NUMBER: "number",
};

export { RULES }