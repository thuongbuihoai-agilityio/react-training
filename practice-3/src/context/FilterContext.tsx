import { createContext, useState } from "react";

export const FilterContext = createContext<any>({});
const MainContext: React.FC<{children: JSX.Element[]}> = ({ children }) => {
  const [filterInput, setFilterInput] = useState({});

  return (
    <FilterContext.Provider value={{ filterInput, setFilterInput }}>
      {children}
    </FilterContext.Provider>
  )
}

export default MainContext;
