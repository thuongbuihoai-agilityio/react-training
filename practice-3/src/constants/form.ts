import { RULES } from "./rules"

const FORM_VALUES = {
  categoryId: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  name: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  price: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  quantity: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  description: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  images: {
    value: "",
    rules: [],
    error: "",
  }
}

export default FORM_VALUES
