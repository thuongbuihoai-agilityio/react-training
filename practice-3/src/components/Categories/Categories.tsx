import useSWR from "swr";
import React, { useContext, useState } from "react";
import { CategoryProps } from "@/types/category";
import { SearchContext } from "@/context/SearchContext";
import { CATEGORIES_URL } from "@/constants/url";
import { get } from "@/helpers/fetchApi";
import "./categories.css";

const Categories: React.FC = () => {
  const { data } = useSWR(CATEGORIES_URL, get);
  const [activeId, setActiveId] = useState("");
  const { setSearchValue } = useContext(SearchContext);
  const handleSearch = (id: string) => (e: React.MouseEvent<HTMLElement>) => {
    const categoryId = { categoryId: e.currentTarget.dataset.index };
    setSearchValue?.(categoryId);
    setActiveId(id);
  };

  const handleDefaultCategory = () => {
    const categoryId = "";
    setSearchValue?.(categoryId);
    setActiveId("");
  };

  function renderCategoryList(data: []) {
    return data?.map((category: CategoryProps) => (
      <li
        data-index={category.id}
        className={`categories__item ${
          activeId === category.id ? "active" : "inactive"
        }`}
        key={category.id}
        onClick={handleSearch(category.id)}
      >
        {category.name}
      </li>
    ));
  }

  return (
    <div data-testid="categories" className="categories">
      <p className="categories__title">What are you looking for here?</p>
      <ul className="categories__list">
        <li
          data-testid="category-item"
          className={`categories__item ${
            activeId === "" ? "active" : "inactive"
          }`}
          onClick={handleDefaultCategory}
        >
          All
        </li>
        {renderCategoryList(data as [])}
      </ul>
    </div>
  );
};

export default Categories;
