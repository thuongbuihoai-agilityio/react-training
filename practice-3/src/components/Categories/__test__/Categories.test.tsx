import { fireEvent, render } from "@testing-library/react";
import { FilterContext } from "@/context/FilterContext";
import Categories from "../Categories";
import "@testing-library/jest-dom";
import { useState } from "react";
import { CATEGORY_MOCKING_LIST } from "@/constants/categories";

const contextValueMock: any = {
  setFilterInput: jest.fn()
};

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

const useCategoriesMock = { categories: CATEGORY_MOCKING_LIST };
  jest.mock("../../../hooks/useCategories.ts", () => ({
    default: jest.fn(() => useCategoriesMock),
}));

describe("Category component", () => {
  beforeEach(()=>{
    (useState as jest.Mock).mockImplementation(jest.requireActual("react").useState);
  })

  test("render category component", () => {
    const { getByTestId } = render(<Categories />);
    const categories = getByTestId("categories");
    expect(categories).toBeInTheDocument();
  });

  test("should filter when click category", () => {
    const { getByTestId } = render(
      <FilterContext.Provider value={contextValueMock}>
        <Categories />
      </FilterContext.Provider>,
    );
    const categoryItem = getByTestId("category-item");
    fireEvent.click(categoryItem);
    expect(contextValueMock.setFilterInput).toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <Categories />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});