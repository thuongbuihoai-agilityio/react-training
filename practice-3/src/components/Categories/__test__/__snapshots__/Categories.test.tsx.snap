// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="categories"
    data-testid="categories"
  >
    <p
      class="categories__title"
    >
      What are you looking for here?
    </p>
    <ul
      class="categories__list"
    >
      <li
        class="categories__item active"
        data-testid="category-item"
      >
        All
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
